### ISSUE TITLE
Pre-submission inquiry: "derivative work" MCMC library

### ISSUE THREAD

**Issue by eigenfoo** [openjournals/joss#824]
I'd like to ask if LittleMCMC is in scope for a JOSS submission: https://github.com/eigenfoo/littlemcmc

It satisfies the [research software](https://joss.readthedocs.io/en/latest/submitting.html#what-we-mean-by-research-software) requirement (for example, similar MCMC packages like [emcee](https://joss.theoj.org/papers/10.21105/joss.01864) and [greta](https://joss.theoj.org/papers/10.21105/joss.01601) have been published in JOSS before).

However, LittleMCMC is pretty transparently based on code from [PyMC3](https://github.com/pymc-devs/pymc3): LittleMCMC's main value proposition is that it isolates PyMC3's robust internal engine (its HMC and NUTS samplers) into a lightweight and standalone library, so that users with custom code (logp functions and their gradients) can use the same internals as PyMC3. In other words, it makes it so that users can "BYOGradient" and still benefit from PyMC3's time-tested samplers. My concern is whether JOSS will accept a "derivative work" like this.

Separately, I'm not sure if LittleMCMC satisfies the [substantial scholarly effort](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort) requirement - specifically, the "number of authors" requirement (it's just me), and the "length of commit history" requirement (very intermittent development for ~ 1 year).

Thanks in advance!

**Comment by jedbrown** [comment]
I'm curious why a stripped-down fork/derivative of PyMC3 is preferable to having PyMC3 expose lower-level interfaces? What would be needed to enable users of LittleMCMC to move to PyMC3?

Also, could you explain what "transparently based on" means?

**Comment by eigenfoo** [comment]
By "transparently based on", I just mean that I don't try to hide this fact! It's clearly stated on [the docs](https://littlemcmc.readthedocs.io/en/latest/), for example.

As to why a stripped-down version is preferable, Bayesian modelling relies on being able to specify a log probability (logp) function and take its gradient (dlogp). PyMC3 relies on Theano to do both of these things, and this abstraction leaks into the samplers - some parts of the sampler assume Theano objects, and it's non-trivial to refactor it.

More importantly, some users will have implemented their logps and dlogps, either for legacy or for performance reasons. Since PyMC3 cannot, for example, accept arbitrary Python callables for the logp and dlogp function, these users would have to either re-implement their model in PyMC3's Theano-based distributions, or write their own sampler - both of which are suboptimal and get in the way of good science.

In short, the ideal LittleMCMC user would be using it not because it is a persuasive alternative to PyMC3, but because the cost of migrating to PyMC3 would be intolerable.

If you'd like more color, I explain a bit more in [this blog post](https://eigenfoo.xyz/littlemcmc/), and there is more context in [this forum post](https://discourse.pymc.io/t/isolate-nuts-into-a-new-library/3974).

**Comment by jedbrown** [comment]
Thanks for your further explanation.

Sounds like @dfm (a JOSS editor) is familiar with what you've done and would be in a good position to comment.

Out of curiosity (not necessarily germane to the scope question), what is stopping PyMC3 from using LittleMCMC?

**Comment by eigenfoo** [comment]
That's a good question! For example, the Turing language (which does probabilistic programming in Julia) does exactly that with [`AdvancedHMC.jl`](https://github.com/TuringLang/AdvancedHMC.jl).

I can't think of any compelling reasons why PyMC3 shouldn't use LittleMCMC, once LittleMCMC is bug-free and reaches feature parity to PyMC3's HMC/NUTS sampler. One wrinkle would be that PyMC3 has a lot more samplers than just HMC/NUTS (a full list can be found in the [`step_methods` module](https://github.com/pymc-devs/pymc3/tree/master/pymc3/step_methods)), and it would be somewhat awkward to expose lower-level interfaces to only _two_ of the samplers.

Also hello @dfm, small world! Glad to see you here 😃

**Comment by dfm** [comment]
I'll definitely have a conflict-of-interest on this one because George and I have spoken a lot about this. I certainly think that this would be within scope, but I would have some questions about authorship: Depending on how much of this is a fork vs. a reimplementation, I think that it might make sense for some PyMC3 devs to be invited as co-authors. If this is seen as an extension of the PyMC3 project then I think the authorship and length of the commit history wouldn't be an issue, but perhaps that's not the way you see it @eigenfoo?

**Comment by eigenfoo** [comment]
I'd definitely consider this a fork of (a subset of) PyMC3, rather than a reimplementation. I guess I was referring earlier to the effort it took me to turn PyMC3 into LittleMCMC, and I wasn't thinking about the effort that it took to write the samplers in the first place 😅

Let me reach out to some PyMC3 developers if they'd be interested - I'll come back and tag them here. Thanks!

**Comment by arfon** [comment]
Closing as stale.
